src/com/sun/jna/win32/W32StringUtil.java

```diff
@@ -0,0 +1,390 @@
+ */
+package com.sun.jna.win32;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import com.sun.jna.Memory;
+import com.sun.jna.Native;
+import com.sun.jna.Pointer;
+
+/**
+ * Native string methods for Windows.
+ *
+ * @author Torbj&ouml;rn Svensson, azoff[at]svenskalinuxforeningen.se
+ */
+public abstract class W32StringUtil {
+    /**
+     * Returns true for unicode API.
+     *
+     * @return True for unicode API
+     */
+    public static boolean isAPITypeWide() {
+        return W32APITypeMapper.DEFAULT == W32APITypeMapper.UNICODE;
+    }
+
+    /**
+     * Return the width of TCHAR in C
+     *
+     * @return The width of TCHAR in C
+     */
+    public static int getCharWidth() {
+        if (isAPITypeWide()) {
+            return Native.WCHAR_SIZE;
+        }
+        return 1;
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param pointer The buffer
+     * @return The string
+     * @see #toString(Pointer)
+     * @see #toString(Pointer, int)
+     * @see #toString(Pointer, int, int)
+     */
+    public static String toString(Pointer pointer) {
+        return toString(pointer, 0);
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param pointer The buffer
+     * @param offsetInCharacters The offset in characters
+     * @return The string
+     * @see #toString(Pointer)
+     * @see #toString(Pointer, int, int)
+     */
+    public static String toString(Pointer pointer, int offsetInCharacters) {
+        if (isAPITypeWide()) {
+            return pointer.getWideString(offsetInCharacters * getCharWidth());
+        } else {
+            return pointer.getString(offsetInCharacters);
+        }
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param pointer The buffer
+     * @param offsetInCharacters The offset in characters
+     * @param lengthInCharacters The length of the string in characters
+     * @return The string
+     * @see #toString(Pointer)
+     * @see #toString(Pointer, int)
+     */
+    public static String toString(Pointer pointer, int offsetInCharacters, int lengthInCharacters) {
+        byte[] bytes = pointer.getByteArray(offsetInCharacters * getCharWidth(), lengthInCharacters * getCharWidth());
+        return toString(bytes, 0);
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param buffer The buffer
+     * @return The string
+     * @see #toString(byte[], long)
+     */
+    public static String toString(byte[] buffer) {
+        return toString(buffer, 0);
+    }
+
+    /**
+     * Convert buffer to string
+     *
+     * @param buffer The buffer
+     * @param offsetInCharacters The offset in characters
+     * @return The string
+     * @see #toString(byte[])
+     */
+    public static String toString(byte[] buffer, long offsetInCharacters) {
+        Memory pointer = new Memory(buffer.length + 1 * getCharWidth());
+        pointer.write(0, buffer, 0, buffer.length);
```

Comment:
So to get a string we transition to native multiple times - why not pass the pointer here and read directly? This hammers the heap _and_ native memory.

Reply:
How do I use the byte buffer as source so that the ptr.getString(0)/ptr.getWideString(0) will read from it?
There is one call path that does a conversion from pointer to byte[] and then read from that and it's the User32Util.loadString method. The reason is that LoadString does not guarantee that the string is null terminated when using it with a null buffer (to get the read only string resource itself and not a copy) and still be able to use getString/getWideString methods to get the correct encoded string.

If there is a way to get access to the getString/getWideString without having it backed by native memory, please enlighten me.

Reply:
Constructing a string with the native encoding should replace `getString`, while (on windows) constructing a new string with `UTF-16` encoding from the raw bytes will yield get result of `getWideString`. At least to my knowledge the wide variants of the functions either work on `UCS-2` or `UTF-16`. As `UTF-16` is a superset of `UCS-2` deconding as `UTF_16` should always work.

Reply:
> Constructing a string with the native encoding should replace `getString`, while (on windows) constructing a new string with `UTF-16` encoding from the raw bytes will yield get result of `getWideString`. At least to my knowledge the wide variants of the functions either work on `UCS-2` or `UTF-16`. As `UTF-16` is a superset of `UCS-2` deconding as `UTF_16` should always work.

I've read the reply several times now and am having trouble following you here... 
Do you mean that I should simply do something like this?
```
public static String toString(byte[] buffer, long offsetInCharacters) {
  if (isAPITypeWide()) {
    return new String(buffer, offsetInBytes, lengthInBytes, StandardCharsets.UTF-16);
  } else {
    return Native.toString(buffer);
  }
}
```

Reply:
For the non-wide variant the `String` constructor with the native encoding should be good.